plugin/storage/badger/factory.go

```diff
@@ -190,3 +205,56 @@ func (f *Factory) maintenance() {
 		}
 	}
 }
+
+func (f *Factory) metricsCopier() {
+	metricsTicker := time.NewTicker(f.Options.primary.MetricsUpdateInterval)
+	defer metricsTicker.Stop()
+	for {
+		select {
+		case <-f.maintenanceDone:
+			return
+		case <-metricsTicker.C:
+			expvar.Do(func(kv expvar.KeyValue) {
+				if strings.HasPrefix(kv.Key, "badger") {
+					if intVal, ok := kv.Value.(*expvar.Int); ok {
+						if g, found := f.metrics.badgerMetrics[kv.Key]; found {
+							g.Update(intVal.Value())
+						}
+					} else if mapVal, ok := kv.Value.(*expvar.Map); ok {
+						mapVal.Do(func(innerKv expvar.KeyValue) {
+							// The metrics we're interested in have only a single inner key (dynamic name)
+							// and we're only interested in its value
+							if intVal, ok := innerKv.Value.(*expvar.Int); ok {
+								if g, found := f.metrics.badgerMetrics[kv.Key]; found {
+									g.Update(intVal.Value())
+								}
+							}
+						})
+					}
+				}
+			})
+		}
+	}
+}
+
+func (f *Factory) registerBadgerExpvarMetrics(metricsFactory metrics.Factory) {
+	f.metrics.badgerMetrics = make(map[string]metrics.Gauge)
+
+	expvar.Do(func(kv expvar.KeyValue) {
+		if strings.HasPrefix(kv.Key, "badger") {
+			if _, ok := kv.Value.(*expvar.Int); ok {
+				g := metricsFactory.Gauge(metrics.Options{Name: kv.Key})
+				f.metrics.badgerMetrics[kv.Key] = g
+			} else if mapVal, ok := kv.Value.(*expvar.Map); ok {
+				mapVal.Do(func(innerKv expvar.KeyValue) {
```

Comment:
wouldn't this map always be empty at this point?

Reply:
No, since badger.Open() is called before the metrics are fetched. And badger.Open() registers the expvars (it calculates the initial sizes during Open()).